### PR TITLE
Refactor node sources

### PR DESCRIPTION
**Pull request by einarmo**
Major refactor to node sources.

The problem has been that each node source effectively needs to replicate the whole node ingestion logic. The common parts were included in an abstract base class, but that really just made things harder to understand, as it required anyone working on a source to understand what the base class was doing at all times.

This more or less flips the dependency, making the base class (now `NodeHierarchyBuilder`, and not a base class, a wrapper) call methods on sources as needed.

This also abstracts out the reading of types from the type manager, making it use the NodeSetSource when that is configured to contain type information, which before worked through a side channel.

There is certainly a lot more to be done, but this is a start.

### CODE REVIEW

File: Extractor/Pushers/FDM/NodeTrimmer.cs [hunk 1]
```diff
@@ -62,7 +63,7 @@ private void TraverseNode(List<BaseUANode> result, List<UAReference> refResult,
 
             if (!bySource.Any() && !byTarget.Any())
             {
-                log.LogWarning("Orphaned node: {Name} {Id}", node.Name, node.Id);
+                throw new InvalidOperationException($"Orphaned node: {node.Name} {node.Id}");
```
[the remark this answers is not in the document]
Comment (einarmo) [reply Extractor/Pushers/FDM/NodeTrimmer.cs:66]: Not really. We do catch anything going on down here, but this is a bug, so it isn't really something we handle gracefully.

File: Extractor/UAExtractor.cs [hunk 2]
```diff
@@ -590,102 +592,86 @@ public async Task WaitForSubscriptions(int timeout = 100)
...
+            if (result.ShouldBackgroundBrowse && initial)
             {
-                var tasks = await MapUAToDestinations(result);
-                if (initial && !readFromOpc && Config.Source.AltSourceBackgroundBrowse)
+                tasks = tasks.Append(async token =>
                 {
-                    tasks = tasks.Append(async token =>
+                    builder = MakeBuilder(uaSource);
+                    result = await builder.LoadNodeHierarchy(true, token);
+                    var nextTasks = await MapUAToDestinations(result);
```
[the remark this answers is not in the document]
Comment (einarmo) [reply Extractor/UAExtractor.cs:666]: Oof, yes it should. Nice catch

[the remark this answers is not in the document]
Comment (einarmo) [reply Extractor/UAExtractor.cs:666]: Maybe it should even just wait on them here. No reason for this task to terminate.

File: Extractor/Pushers/FDM/NodeHierarchy.cs [hunk 3]
```diff
@@ -20,7 +20,7 @@ public NodeHierarchy(IEnumerable<UAReference> references, IEnumerable<BaseUANode
             ReferencesByTargetId = references.Where(rf => rf.IsForward).GroupBy(rf => rf.Target.Id).ToDictionary(group => group.Key, group => (IEnumerable<UAReference>)group);
             ReferencesBySourceId = references.Where(rf => rf.IsForward).GroupBy(rf => rf.Source.Id).ToDictionary(group => group.Key, group => (IEnumerable<UAReference>)group);
             NodeMap = nodes.ToDictionary(node => node.Id);
-            KnownTypeDefinitions = new HashSet<NodeId>(nodes.Where(node => node.Id.NamespaceIndex > 0 || !IsChildOfType(node)).SelectNonNull(s => s.TypeDefinition));
+            KnownTypeDefinitions = new HashSet<NodeId>(nodes.Where(node => node.Id.NamespaceIndex > 0 || !node.IsChildOfType).SelectNonNull(s => s.TypeDefinition));
         }
 
         private bool IsChildOfType(BaseUANode node)
```
[the remark this answers is not in the document]
Comment (einarmo) [reply Extractor/Pushers/FDM/NodeHierarchy.cs:26]: Yep, nice catch. Now that we have generalized all the node hierarchy construction in the main part of the extractor itself, this should be unnecessary.

There may be bugs related to this, but it should be fine for the time being, this part is getting changed after all.

**Review by ozangoktan — APPROVED** [review]

**Pull request merged**